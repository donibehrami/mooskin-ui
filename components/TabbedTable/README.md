# MooSkin UI - TabbedTable Component

The MooSkin TabbedTable Component is like a tabbed window, with as many tabs as you like.

___

### Usage

To start using the TabbedTable Component first you have to Import its two main components

```
Import {TabbedTable, TabTable} from 'mooskin';
```
or
```
import TabbedTable, {TabTable} from 'mooskin/lib/TabbedTable';
```

And then you can simply start using it by typing

```
<TabbedTable>
    <TabTable title="Title 1" active>
        <thead>
            <tr>
                <th>th 1</th>
                <th>th 2</th>
            </tr>
        </thead>
        <tbody>
            <tr>
                <td>td 1</td>
                <td>td 2</td>
            </tr>
            <tr>
                <td>td 1</td>
                <td>td 2</td>
            </tr>
        </tbody>
    </TabTable>
    <TabTable title="Title 2">
        <thead>
            <tr>
                <th>th 1</th>
                <th>th 2</th>
            </tr>
        </thead>
        <tbody>
            <tr>
                <td>td 1</td>
                <td>td 2</td>
            </tr>
            <tr>
                <td>td 1</td>
                <td>td 2</td>
            </tr>
        </tbody>
    </TabTable>
</TabbedTable>
```


Individual TabTable components have a required 'title' prop which will be displayed in the TabTable header, the rest are optional. The content of the TabTable should be table contents starting from `thead` or `tbody`.

### Examples

Can be used without importing the TabTable component with shorthand dot(.) notation.

```
Import {TabbedTable} from 'mooskin';

<TabbedTable>
    <TabbedTable.TabTable title="Title 1" iconClass="fa fa-icon">
        <div style={{...style, backgroundColor: '#5CCDDF'}}>
            Content 1
        </div>
    </TabbedTable.TabTable>
    <TabbedTable.TabTable title="Title 2" materialIcon="home">
        <div style={{...style, backgroundColor: '#F48770'}}>
            Content 2
        </div>
    </TabbedTable.TabTable>
</TabbedTable>
```

<div class="playground-doc">

## Supported attributes for TabbedTable

* `id` - id of the element
* `className` - css class
* `style` - extra styles for the main container

## Supported attributes for TabTable 

* `title` - (required) sets the title for the tab
* `active` - sets which TabTable is active, the first is active by default
* `value` - an optional numerical value displayed on the header
* `info` - another optional information displayed on the header
* `style` -  extra styles 
* `iconClass` - a class that will be added before the title to act as an icon, for example a font awesome class
* `materialIcon` - name of the material icon to be added before the title (material icons are provided out of the box)

</div>

Allthough these attributes are supported, most of them are optional.

___

#### For more

[MooSkin-UI](https://github.com/moosend/mooskin-ui)