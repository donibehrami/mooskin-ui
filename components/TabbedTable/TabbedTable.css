@import url('https://fonts.googleapis.com/css?family=Convergence');
@import '../_utils/globals/variables.css';
@import '../_utils/globals/screens.css';

.container {
    width: 100%;
}

.heading {
    width:100%;
    height:95px;
    border-bottom:10px solid #efefef;
    display:flex;
    flex-direction: row;
    border-top-left-radius: 5px;
    border-top-right-radius: 5px;
}

.header {
    flex-grow:1;
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
    cursor: pointer;
    transition: all 400ms ease;
    border-top-left-radius: 5px;
    border-top-right-radius: 5px;
}

.header:hover > .title{
    color: var(--moosendBlue);
}

.title{
    color: #6b6b6b;
    font-weight: 600;
    font-size: .92857142857rem;
    /*font-size: 18px;*/
}

.image{
    width: 40px;
}

.value{
    font-size: 1rem;
    font-weight: 800;
    color: #636363;
}

.info{
    font-size: .78571428571rem;
    color: #aaa;
}

.activeHeader {
    background: var(--inputBackground);
}

.inactiveHeader {
    background: #fff;
}

.inactiveHeader:hover {
    background: rgba(245, 245, 245, 0.5);
}

.table{
    width: 100%;
    border-collapse: collapse;
    border: 1px solid #eaeaea;
}

.table thead tr th{
    background-color: #efefef;
    padding-top: 8px;
    padding-bottom: 8px;
    color: #aaaaaa;
    font-size: .8rem;
    border-bottom: 1px solid #d4d4d4;
    font-family: Arial, Helvetica, sans-serif;
    font-weight: normal;
}

.table tbody tr td{
    text-align: center;
    vertical-align: middle;
    padding-top: 20px;
    padding-bottom: 20px;
    font-size: 1rem;
    border-bottom: 1px solid #eaeaea;
    color:#6b6b6b;
}

.visible {
    display: table;
}

.invisible {
    display: none;
}

@media (--deviceSmall){
    .table thead tr th{
        font-size: 2.5vw;
    }

    .table tbody tr td{
        font-size: 2.5vw;
    }
}

@media (--deviceXsmall){
    .table thead tr th{
        font-size: 2.5vw;
    }

    .table tbody tr td{
        font-size: 2.5vw;
    }
}