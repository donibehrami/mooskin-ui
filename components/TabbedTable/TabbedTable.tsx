import * as React from 'react';

import styles from './TabbedTable.css';

// import {Table} from '../index';
// import ITableProps from '../Table';

export interface ITabbedTableProps {

    /** override id */
    id?: string;

    /** container class */
    className?: string;

    /** override styles */
    style?: React.CSSProperties;

    /** children here can only be Tab elements */
    children?: Array<React.ReactElement<ITabTableProps>> | React.ReactElement<ITabTableProps>;
}

export interface ITabTableProps{

    /** override Table id */
    id?: string;

    /** title */
    title: string;

    /** further information to be displayed on the header */
    info?: string;

    /** value displayed on the tab header */
    headerValue?: number;

    /** image displayed within the tab header */
    image?: string;

    /** Table class */
    className?: string;

    /** override Table styles */
    style?: React.CSSProperties;

    /** wether this table is active or not */
    active?: boolean;

    /** onClick callback function when a tab header is clicked */
    onClick?: (e: React.MouseEvent<HTMLElement>) => void;
}

export interface IHeaderProps {
    value?: number;
    info?: string;
    href?: string;
    image?: string;
    title: string;
    active: boolean;
    onClick: (e: React.MouseEvent<HTMLElement>) => void;
}

export interface ITabbedTableState {
    activeTable?: number;
}

export default class TabbedTable extends React.Component<ITabbedTableProps, ITabbedTableState>{

    public static defaultProps = {
        className: '',
        style: {}
    };

    public static TabTable: React.StatelessComponent<ITabbedTableProps>;

    constructor(props: any){
        super(props);

        this.state = {
            activeTable: this.getActiveTable()
        };
    }

    public render(){

        const {headers, tables} = this.makeContent();

        return(
            <div className={`tabbed-table-component ${styles.container}`}>
                <div className={styles.heading}>
                    {headers}
                </div>
                <div className={styles.tables}>
                    {tables}
                </div>
            </div>
        );
    }

    private makeContent(){

        const headers: Array<React.ReactElement<IHeaderProps>> = [];
        const tables: Array<React.ReactElement<ITabTableProps>> = [];

        React.Children.forEach(this.props.children, (child, index) => {
            if (React.isValidElement<ITabTableProps>(child)){

                headers.push(
                    <Header
                        key={index}
                        value={child.props.headerValue}
                        title={child.props.title}
                        info={child.props.info}
                        image={child.props.image}
                        active={this.state.activeTable === index}
                        onClick={this.onClickHeader(index, child)}
                    />
                );

                const extraProps: Partial<ITabTableProps & {key: number}> = {
                    active: this.state.activeTable === index,
                    key: index
                };

                tables.push(React.cloneElement(child, extraProps));

            }else{
                throw new Error('<TabbedTable> element only accepts <TabTable> elements as children');
            }
        });

        return {headers, tables};
    }

    private onClickHeader = (tabIndex: number, table: React.ReactElement<ITabTableProps>) => {
        return (e: React.MouseEvent<HTMLElement>) => {
            this.setState({activeTable: tabIndex});
            table.props.onClick && table.props.onClick(e);
        };
    }

    private getActiveTable(): number {
        let activeTable = 0;
        const childrenArray = React.Children.toArray(this.props.children);

        for (const [index, value] of childrenArray.entries()){
            if (React.isValidElement<ITabTableProps>(value) && value.props.active){
                activeTable = index;
            }
        }

        return activeTable;
    }
}

export const TabTable: React.StatelessComponent<ITabTableProps> = (props) => {

    const displayClass = !props.active ? styles.invisible : styles.visible;

    return (
            <div className={`tab-table-component`}>
                <table {...this.props} className={`${styles.table} ${props.className} ${displayClass}`}>
                    {props.children}
                </table>
            </div>
        );
};

export const Header: React.StatelessComponent<IHeaderProps> = (props) => {

    const activeTab = props.active ? styles.activeHeader : styles.inactiveHeader;

    return (
        <div className={`tab-header ${styles.header} ${styles.anchor} ${activeTab}`} onClick={props.onClick}>
            <span className={styles.title}>{props.title}</span>
            <img src={props.image} className={styles.image}/>
            <span className={styles.value}>{props.value}</span>
            <span className={styles.info}>{props.info}</span>
        </div>
    );
};
