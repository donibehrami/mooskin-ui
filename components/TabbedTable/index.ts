import TabbedTable from './TabbedTable';

export default TabbedTable;
