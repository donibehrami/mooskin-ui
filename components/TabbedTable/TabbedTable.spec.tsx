import * as React from 'react';
import TabbedTable, {TabTable} from './TabbedTable';

import {  mount, shallow } from 'enzyme';

describe('TabbedTable', () => {

    test('renders correctly', () => {
        const func = jest.fn();

        const tree = shallow(
            <TabbedTable className="myClass" style={{color: 'red'}}>
                <TabTable title={'Table title'} href="/reports" info="some info">
                    <thead>
                        <tr>
                            <th>th 1</th>
                            <th>th 2</th>
                            <th>th 3</th>
                        </tr>
                    </thead>
                    <tbody>
                        <tr>
                            <td>td 1</td>
                            <td>td 2</td>
                            <td>td 3</td>
                        </tr>
                        <tr>
                            <td>td 1</td>
                            <td>td 2</td>
                            <td>td 3</td>
                        </tr>
                    </tbody>
                </TabTable>
                <TabTable title={'Table title 2'} href="/maillist" info="some info">
                    <thead>
                        <tr>
                            <th>th 1</th>
                            <th>th 2</th>
                            <th>th 3</th>
                        </tr>
                    </thead>
                    <tbody>
                        <tr>
                            <td>td 1</td>
                            <td>td 2</td>
                            <td>td 3</td>
                        </tr>
                        <tr>
                            <td>td 1</td>
                            <td>td 2</td>
                            <td>td 3</td>
                        </tr>
                    </tbody>
                </TabTable>
            </TabbedTable>
        );
        expect(tree).toMatchSnapshot();
    });

    test('renders 2 TabTables with children', () => {
        const func = jest.fn();

        const component = shallow(
            <TabbedTable className="myClass" style={{color: 'red'}}>
                <TabTable title={'Table title'} href="/reports" info="some info">
                    <thead>
                        <tr>
                            <th>th 1</th>
                            <th>th 2</th>
                            <th>th 3</th>
                        </tr>
                    </thead>
                    <tbody>
                        <tr>
                            <td>td 1</td>
                            <td>td 2</td>
                            <td>td 3</td>
                        </tr>
                        <tr>
                            <td>td 1</td>
                            <td>td 2</td>
                            <td>td 3</td>
                        </tr>
                    </tbody>
                </TabTable>
                <TabTable title={'Table title 2'} href="/maillist" info="some info">
                    <thead>
                        <tr>
                            <th>th 1</th>
                            <th>th 2</th>
                            <th>th 3</th>
                        </tr>
                    </thead>
                    <tbody>
                        <tr>
                            <td>td 1</td>
                            <td>td 2</td>
                            <td>td 3</td>
                        </tr>
                        <tr>
                            <td>td 1</td>
                            <td>td 2</td>
                            <td>td 3</td>
                        </tr>
                    </tbody>
                </TabTable>
            </TabbedTable>
        );

        expect(component.find(TabTable).length).toEqual(2);
        expect(component.find('td').length).toEqual(12);

        expect(component.find(TabTable).first().prop('title')).toEqual('Table title');
        expect(component.find(TabTable).first().prop('href')).toEqual('/reports');
        expect(component.find(TabTable).first().prop('info')).toEqual('some info');
    });

});
