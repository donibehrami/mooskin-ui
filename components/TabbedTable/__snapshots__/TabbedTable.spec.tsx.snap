// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TabbedTable renders correctly 1`] = `
<div
  className="tabbed-table-component container"
>
  <div
    className="heading"
  >
    <Header
      active={true}
      href="/reports"
      info="some info"
      onClick={[Function]}
      title="Table title"
    />
    <Header
      active={false}
      href="/maillist"
      info="some info"
      onClick={[Function]}
      title="Table title 2"
    />
  </div>
  <div
    className="tables"
  >
    <TabTable
      active={true}
      href="/reports"
      info="some info"
      title="Table title"
    >
      <thead>
        <tr>
          <th>
            th 1
          </th>
          <th>
            th 2
          </th>
          <th>
            th 3
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            td 1
          </td>
          <td>
            td 2
          </td>
          <td>
            td 3
          </td>
        </tr>
        <tr>
          <td>
            td 1
          </td>
          <td>
            td 2
          </td>
          <td>
            td 3
          </td>
        </tr>
      </tbody>
    </TabTable>
    <TabTable
      active={false}
      href="/maillist"
      info="some info"
      title="Table title 2"
    >
      <thead>
        <tr>
          <th>
            th 1
          </th>
          <th>
            th 2
          </th>
          <th>
            th 3
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            td 1
          </td>
          <td>
            td 2
          </td>
          <td>
            td 3
          </td>
        </tr>
        <tr>
          <td>
            td 1
          </td>
          <td>
            td 2
          </td>
          <td>
            td 3
          </td>
        </tr>
      </tbody>
    </TabTable>
  </div>
</div>
`;
